File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack =  []
        checkClosed = {")" : "(" , "]" : "[" , "}" : "{", }
        for i in s:
            if i in checkClosed:
                if stack and stack[-1] == checkClosed[i] : 
                    stack.pop()
                else :
                    return False
            else :
                stack.append(i)
                
        return True if not stack else  False
        
        

File: 0020-valid-parentheses/0020-valid-parentheses.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        HashMap<Character, Character> checkClosing = new HashMap<>();
        checkClosing.put(')', '(');
        checkClosing.put(']', '[');
        checkClosing.put('}', '{');
        
        for (char i : s.toCharArray()) {
            if (checkClosing.containsKey(i)) {
                if (!stack.isEmpty() && stack.peek() == checkClosing.get(i)) {
                    stack.pop();
                } else {
                    return false;
                }
            } else {
                stack.push(i);
            }
        }
        
        return stack.isEmpty();
    }
}
